---
title: "Preprocessing_ITS_Part2"
author: "Yupawadee Galasong"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

### Load Packages and Functions

```{r load packages}
pacman::p_load(tidyverse, phyloseq,PERFect, install = FALSE)
```

### Load raw phyloseq data from DADA2 workflow

```{r load phyloseq obj}

load(file = "Data_ITS/raw_physeq.RData")
```

### Assessing mock community sequencing results

```{r 16S mock only, eval=FALSE, include=FALSE}
library(devtools)
install_github("microsud/chkMocks")

```

The mock communities we used is ZymoBIOMICS Microbial Community DNA Standard II (Log Distribution) which has the theoretical composition of genomic DNA from *Cryptococcus neoformans* (0.00089%) and *Saccharomyces cerevisiae* (0.89%). The remaining genomic DNA is from bacteria, which we will examine more closely & more systematically using the package `r chkMocks`. For the ITS sequencing run, we will mainly evaluate the mock community for the presence/absence of the taxa.

```{r}
# Select only mock communities
mock_physeq <- prune_samples(raw_physeq, samples = c("ITSMock1","ITSMock2","ITSMock3"))
# Select only ASVs of abundance > 0
mock_physeq <- prune_taxa(taxa_sums(mock_physeq)>0,mock_physeq)

# Genus and Species of ASVs detected in mock communites
mock_physeq@tax_table[,c("Genus","Species")]

# Calculate the ratio of Saccharomyces: Cryptococcus in each mock samples
mock_physeq@otu_table["ASV_7"]/
  ((mock_physeq@otu_table["ASV_347"]+
      (mock_physeq@otu_table["ASV_377"])))

```

As we can see, both *Cryptococcus neoformans* and *Saccharomyces sp.* are present in the mock community samples that we sequenced. On average, the number of ASVs assigned to *Saccharomyces sp.* is approximately 1000 times more than ASVs assigned to *C. neoformans*, with ITSMock1 being the only exception. Overall, our sequencing run did not deviate too much from the theoretical composition of the mock community.

### Removing unwanted ASVs

The goal of this pre-processing is to remove the following ASVs from our samples: 1. Mitochondria ASVs 2. Chloroplast ASVs 3a. ASVs found in controls 3b. Control samples 4. ASVs of mock community 5. ASVs undefined at Phylum level 6. ASVs of 0 abundance

Other than ASVs found in control samples, we can remove unwanted ASVs easily using subset\_ or prune\_ commands in phyloseq.

```{r}
raw_physeq_cleaned <- raw_physeq %>% 
  # Remove mitochondria
  subset_taxa(Family != "Mitochondria" | is.na(Family)) %>%
  # Remove chloroplast
  subset_taxa(Order != "Chloroplast" | is.na(Order)) %>%
  # Remove the mock samples
  prune_samples((sample_names(.) != sample_names(mock_physeq)),.) %>%
  # Remove ASVs that have abundance = 0 across samples
  prune_taxa(taxa_sums(.)>0,.) %>%
  # Remove ASVs undefined at Phylum level
  subset_taxa(!is.na(Phylum))

```

Next, let's look at the PCR negative control and DNA extraction negative control samples.

```{r}
ctrl_physeq <- raw_physeq %>% 
  # Subset only control samples
  subset_samples(Sample_or_Control == "Control Sample") %>%
  # Exclude mock samples
  prune_samples(sample_names(.) != sample_names(mock_physeq),.) 

# Confirm that only PCR negative controls and extraction negative controls are 
# in this phyloseq object
sample_names(ctrl_physeq)

# Remove taxa with abundance = 0 from the control samples
ctrl_physeq <- prune_taxa(taxa_sums(ctrl_physeq)>0,ctrl_physeq)

# Genus and Species of ASVs detected in control samples
ctrl_physeq@tax_table[,c("Genus","Species")]
```

It turned out that some taxa are present in both negative controls and true samples. We will use a statistical package `r decontam` to determine which contaminant sequences should be removed.

#### Removal of contaminant sequences with decontam

Following the tutorial <https://benjjneb.github.io/decontam/vignettes/decontam_intro.html>

```{r}
library(devtools)
#devtools::install_github("benjjneb/decontam")
library(decontam)
```

##### Inspect library sizes

```{r}
df <- as.data.frame(sample_data(raw_physeq_cleaned))
df$LibrarySize <- sample_sums(raw_physeq_cleaned)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data = df, aes(x = Index, y = LibrarySize, color = Sample_or_Control))+
  geom_point()

```

There is an outlier - a negative control samples with \~30,000 reads.

##### Identify contaminants - Prevalence method

"In our phyloseq object, `"Sample_or_Control"` is the sample variable that holds the negative control information. We'll summarize that data as a logical variable, with TRUE for control samples, as that is the form required by `isContaminant`."

```{r}
sample_data(raw_physeq_cleaned)$is.neg <- 
  sample_data(raw_physeq_cleaned)$Sample_or_Control == "Control Sample"

# Add 'Batch' information for contaminant identification
raw_physeq_cleaned@sam_data["ITSExtNeg1","Trial"] = 1
raw_physeq_cleaned@sam_data["ITSExtNeg2","Trial"] = 2
raw_physeq_cleaned@sam_data["ITSExtNeg3","Trial"] = 3
raw_physeq_cleaned@sam_data["ITSPCRNeg1","Trial"] = 1
raw_physeq_cleaned@sam_data["ITSPCRNeg2","Trial"] = 2
raw_physeq_cleaned@sam_data["ITSPCRNeg3","Trial"] = 3

# threshold = The probability threshold below which (strictly less than) the null-hypothesis (not a contaminant) should be rejected in favor of the alternate hypothesis (contaminant)
# Increasing the threshold generally results in more sequences being flagged as contaminant.
contamdf.prev <- isContaminant(raw_physeq_cleaned, method = "prevalence",
                               neg = "is.neg", 
                               threshold = 0.5, 
                               batch = sample_data(raw_physeq_cleaned)$Trial)
table(contamdf.prev$contaminant)

# Which sequence is the contaminant?
contam <- subset_taxa(raw_physeq_cleaned, contamdf.prev$contaminant)
taxa_names(contam)
```

##### Plot the number of times the contaminant taxa were observed in samples

```{r}
# Make phyloseq object of presence-absence (pa) in negative controls and true sample
# Present = 1, Absent = 0
pa <- transform_sample_counts(raw_physeq_cleaned, function(abund) 1*(abund>0))
pa.neg <- prune_samples(sample_data(pa)$Sample_or_Control == "Control Sample", pa)
pa.pos <- prune_samples(sample_data(pa)$Sample_or_Control == "True Sample", pa)

# Make data frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos = taxa_sums(pa.pos), pa.neg = taxa_sums(pa.neg),
                    contaminant = contamdf.prev$contaminant)
ggplot(data = df.pa, aes(x = pa.neg, y = pa.pos, color = contaminant))+
  geom_point(alpha = 0.5)+ xlab("Prevalence (Negative Controls)") + 
  ylab("Prevalence (True Samples)")
```

##### Remove contaminant sequences

```{r}

final_physeq <- raw_physeq_cleaned %>% 
  # Remove contaminant sequences
  prune_taxa(!(taxa_names(.) %in% taxa_names(contam)),.) %>%
  # Remove control samples as they have served their purpose
  subset_samples(Sample_or_Control != "Control Sample")
save(final_physeq, file = "Data_ITS/final_physeq.RData")
```
#### Filter rare taxa using PERFect algorithm

The filtering methods for this package are wrapped into two main functions, PERFect_sim() (performing simultaneous filtering) and PERFect_perm() (performing permutation filtering). 
##### Simultaneous filtering
```{r}
#load(file = "Data_ITS/final_physeq.RData")
Counts <- t(final_physeq@otu_table)
dim(Counts)
```
By default, the function PERFect_sim() takes the data table, $X$, as a matrix or data frame, orders it by the taxa abundance, uses 10%, 25% and 50% quantiles for matching the log of DFL to a Skew-Normal distribution and then calculates the p-value for each taxon at the significance level of $\alpha$ = 0.1. The function PERFect_sim() only needs a taxa table as the input, and other parameters are set to default.
```{r}
res_sim <- PERFect_sim(X = Counts)

# The "filtX" object from the result stores the filtered OTU matrix

dim(res_sim$filtX)      

# Signal taxa
sig_asv_sim<- colnames(res_sim$filtX) 
```
Extract and plot p-values for all taxa
```{r}
head(res_sim$pvals)

#plot using the function pvals_Plots().

p <- pvals_Plots(PERFect = res_sim, X = Counts, quantiles = c(0.25, 0.5, 0.8, 0.9), alpha=0.05)

p$plot + ggtitle("Simultanenous Filtering")
```
Alternatively, we can use permutation filtering PERFect_perm() which is more robust than simultaneous filtering. By default, this function generates k = 10000 permutations for each taxa, thus it can be computationally expensive for a large OTU matrix. The package offers user a fast algorithm which employs an unbalanced binary search that optimally finds the cutoff taxon without building the permutation distribution for all taxa. 
##### Permutation filtering
Warning: the 'full' algorithm might take up to an hour.
The 'fast' algorithm might take about 20 mins.
```{r}
res_perm <- PERFect_perm(X = Counts, Order = "pvals", pvals_sim = res_sim, algorithm = "full")
res_perm2 <- PERFect_perm(X = Counts, Order = "pvals", pvals_sim = res_sim, algorithm = "fast", rollmean = FALSE)
p1 <- pvals_Plots(res_perm, Counts)
p1 <- p1$plot + ggtitle("Full Algorithm")
p2 <- pvals_Plots(res_perm2, Counts)
p2 <- p2$plot + ggtitle("Fast Algorithm")
ggpubr::ggarrange(p1,p2,ncol = 2, common.legend = TRUE)
```
The figure above illustrates the plot of permutation PERFect p-values calculated by the full and fast algorithm for the mock2 dataset. Although both methods achieve the similar cutoff taxon, the fast algorithm only calculate 11 out 46 p-values hence is more computationally efficient.

```{r}
dim(res_perm$filtX)   
# signal taxa (full algorithm)   
sig_asv_full <- colnames(res_perm$filtX) 

dim(res_perm2$filtX)
# signal taxa (fast algorithm)
sig_asv_fast <- colnames(res_perm2$filtX)


```
Now we will remove the rare taxa from the phyloseq object.
```{r}
PERFect_physeq <- 
subset_taxa(final_physeq, taxa_names(final_physeq) %in% sig_asv_full)
```
#### Remove singletons
```{r}
PERFect_physeq <-
  prune_taxa(taxa_sums(PERFect_physeq) > 1,
             PERFect_physeq)
```
Now, we have a clean phyloseq object ready for diversity analyses!

