---
title: "Preprocessing_ITS"
author: "Yupawadee Galasong"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Load packages
```{r}
pacman::p_load(dada2, tidyverse, phyloseq, patchwork,Biostrings, install = FALSE)
```
# Install an additional package for cutadapt

```{r}
pacman::p_load(devtools, DelayedArray, GenomeInfoDb, ShortRead, install = FALSE)
install_github("omicsCore/SEQprocess")
library("SEQprocess") # Loads the package
#library(help="SEQprocess") # Lists package info
```

# Set up file path
```{r}
path <- "./Data_ITS"
```

# Renaming files
This part will be skipped for now. We only want to see the overall read quality,
sequencing output, and whether the Nakano primers yield lower mitochondria & chloroplast reads
compared to the universal 16S primers.

# Create variables for the forward and reverse reads
```{r}
#Forward read variables
forward_reads <- list.files(path, pattern = "R1", full.names = TRUE)

#Reverse read variables 
reverse_reads <- list.files(path, pattern = "R2", full.names = TRUE)
```

# Create a list of sample names
In Terminal, use ls command to list the file names and pipe the output to a new
txt file called "samples_ITS.txt".
```{r list sample names}
#Remove _R[1,2].fastq.gz extension to obtain true sample names
samples <- list()
filenames <- scan("samples_ITS.txt",character())
for(name in filenames){
  sample <- gsub("_R[0-9].fastq.gz",'',name)
  if(!sample %in% samples){
  samples <- append(samples,sample)
  }
}
#IMPORTANT: sort the 'samples' vector alphabetically so it matches with how R list the files in the path
samples <- sort(unlist(samples))
```

#Create varibales for filtered forward and reverse reads
```{r}
filtered_forward_reads <- file.path(path, "filtered", paste0(samples,"_R1_filtered.fastq.gz"))

filtered_reverse_reads <- file.path(path, "filtered", paste0(samples,"_R2_filtered.fastq.gz"))
```


#Assess Raw Read Quality
The raw read quality for samples #2 and #9 is poor. These are pasteurized orange juice and pasteurized cider, respectively.
```{r Assess Raw Read Quality}
#Display quality plot for forward read samples
forwardQual10_plot <- plotQualityProfile(forward_reads[1:10])
forwardQual20_plot <- plotQualityProfile(forward_reads[11:20])
forwardQual30_plot <- plotQualityProfile(forward_reads[21:30])
forwardQual40_plot <- plotQualityProfile(forward_reads[31:40])
forwardQual50_plot <- plotQualityProfile(forward_reads[41:50])
forwardQual60_plot <- plotQualityProfile(forward_reads[51:60])
forwardQual70_plot <- plotQualityProfile(forward_reads[61:70])
forwardQual80_plot <- plotQualityProfile(forward_reads[71:80])
forwardQual90_plot <- plotQualityProfile(forward_reads[81:90])
forwardQual100_plot <- plotQualityProfile(forward_reads[91:100])
forwardQual112_plot <- plotQualityProfile(forward_reads[101:112])

#Display quality plot for  reverse read samples
reverseQual10_plot <- plotQualityProfile(reverse_reads[1:10])
reverseQual20_plot <- plotQualityProfile(reverse_reads[11:20])
reverseQual30_plot <- plotQualityProfile(reverse_reads[21:30])
reverseQual47_plot <- plotQualityProfile(reverse_reads[31:47])

```
