---
title: "Preprocessing_ITS"
author: "Yupawadee Galasong"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
  markdown: 
    wrap: 72
---

### Load packages

```{r}
pacman::p_load(dada2, tidyverse, phyloseq, patchwork,Biostrings, install = FALSE)

# Install an additional package for cutadapt
pacman::p_load(devtools, DelayedArray, GenomeInfoDb, ShortRead, install = FALSE)
install_github("omicsCore/SEQprocess")
library("SEQprocess") # Loads the package
#library(help="SEQprocess") # Lists package info
```

### Set up file path

```{r}
path <- "./Data_ITS"
```

### Create variables for the forward and reverse reads

```{r}
#Forward read variables
forward_reads <- list.files(path, pattern = "R1", full.names = TRUE)

#Reverse read variables 
reverse_reads <- list.files(path, pattern = "R2", full.names = TRUE)
```

### Create a list of sample names

In Terminal, use ls command to list the file names and pipe the output
to a new txt file called "samples_ITS.txt".

```{r list sample names}
#Remove _R[1,2].fastq.gz extension to obtain true sample names
samples <- list()
filenames <- scan("samples_ITS.txt",character())
for(name in filenames){
  sample <- gsub("_10471140_L9NR7_R[0-9].fastq.gz",'',name)
  if(!sample %in% samples){
  samples <- append(samples,sample)
  }
}
#IMPORTANT: sort the 'samples' vector alphabetically so it matches with how R list the files in the path
samples <- sort(unlist(samples))
```

### Create variables for filtered forward and reverse reads

```{r}
filtered_forward_reads <- file.path(path, "filtered", paste0(samples,"_R1_filtered.fastq.gz"))

filtered_reverse_reads <- file.path(path, "filtered", paste0(samples,"_R2_filtered.fastq.gz"))
```

### Assess Raw Read Quality

This chunk would take long to run during Rmd knitting.
Therefore, after inspecting the plots, all plotQualityProfile chunks' settings 
were changed to 'show nothing (don't run code)'.

```{r Assess Raw Read Quality, eval=FALSE, include=FALSE}
#Display quality plot for forward read samples
forwardQual10_plot <- plotQualityProfile(forward_reads[1:10])
#Total read count 522954
forwardQual20_plot <- plotQualityProfile(forward_reads[11:20])
#Total read count 482898
forwardQual30_plot <- plotQualityProfile(forward_reads[21:30])
#Total read count 378849
forwardQual40_plot <- plotQualityProfile(forward_reads[31:40])
#Total read count 404119
forwardQual50_plot <- plotQualityProfile(forward_reads[41:50])
#Total read count 396690
forwardQual60_plot <- plotQualityProfile(forward_reads[51:60])
#Total read count 469516
forwardQual70_plot <- plotQualityProfile(forward_reads[61:70])
#Total read count 493893
forwardQual80_plot <- plotQualityProfile(forward_reads[71:80])
#Total read count 437201
forwardQual90_plot <- plotQualityProfile(forward_reads[81:90])
#Total read count 435716
forwardQual100_plot <- plotQualityProfile(forward_reads[91:100])
#Total read count 455482
forwardQual110_plot <- plotQualityProfile(forward_reads[101:110])
#Total read count 427092
forwardQual117_plot <- plotQualityProfile(forward_reads[111:117])
#Total read count 211660
#Total count of forward reads generated = 5,116,070

#Display quality plot for  reverse read samples
reverseQual10_plot <- plotQualityProfile(reverse_reads[1:10])
reverseQual20_plot <- plotQualityProfile(reverse_reads[11:20])
reverseQual30_plot <- plotQualityProfile(reverse_reads[21:30])
reverseQual40_plot <- plotQualityProfile(reverse_reads[31:40])
reverseQual50_plot <- plotQualityProfile(reverse_reads[41:50])
reverseQual60_plot <- plotQualityProfile(reverse_reads[51:60])
reverseQual70_plot <- plotQualityProfile(reverse_reads[61:70])
reverseQual80_plot <- plotQualityProfile(reverse_reads[71:80])
reverseQual90_plot <- plotQualityProfile(reverse_reads[81:90])
reverseQual100_plot <- plotQualityProfile(reverse_reads[91:100])
reverseQual110_plot <- plotQualityProfile(reverse_reads[101:110])
reverseQual117_plot <- plotQualityProfile(reverse_reads[111:117])

#Total count of forward reads generated = 5,116,070
#Total count of ITS reads generated = 2 x 5,116,070 = 10,232,140
#Total count of 16S (subset) reads generated = 2 x 230,339 = 460,678
#Total bp output = 250 bp x 10,692,818 = 2.673 Gbp
#This output is much lower than the minimum output ~7.5 Gbp promised by Illumina

```

List of bad quality samples

ITS9 (forward & reverse), ITS10 (forward & reverse), ITS11 (forward &
reverse), ITS12 (forward & reverse), ITS26 (forward & reverse), ITS27
(forward & reverse), ITS 29 (reverse), ITSExtNeg1, ITSExtNeg3
,ITSPCRNeg1, ITSPCRNeg2, ITSPCRNeg3

### Remove primers and adapters with cutadapt

```{r primers info}
fwd_primer <- "AACTTTYRRCAAYGGATCWCT"
rev_primer <- "AGCCTCCGCTTATTGATATGCTTAART"
rev_rc <- toString(reverseComplement(DNAString(rev_primer)))
fwd_rc <- toString(reverseComplement(DNAString(fwd_primer)))

```

#### Some additional steps before primer removal

Following this tutorial:
<https://benjjneb.github.io/dada2/ITS_workflow.html>

```{r primer orientation}
allOrients <- function(primer) {
    # Create all orientations of the input sequence
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
    orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna), 
        RevComp = reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(fwd_primer)
REV.orients <- allOrients(rev_primer)
FWD.orients
```

#### Pre-filtering N's from raw reads & Count primers
```{r filter Ns from raw seqs}
forward_reads.filtN <- file.path(path, "filtN", basename(forward_reads)) # Put N-filterd files in filtN/ subdirectory
reverse_reads.filtN <- file.path(path, "filtN", basename(reverse_reads))
filterAndTrim(forward_reads, forward_reads.filtN, reverse_reads, reverse_reads.filtN, maxN = 0, multithread = TRUE)
primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = forward_reads.filtN[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = reverse_reads.filtN[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = forward_reads.filtN[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = reverse_reads.filtN[[1]]))
```
#### Cutadapt
```{r cutadapt}
cutadapt <- "/home/yg225/cutadapt-venv/bin/cutadapt"
path.cut <- file.path(path, "cutadapt")
if(!dir.exists(path.cut)) dir.create(path.cut)
forward_reads.cut <- file.path(path.cut, basename(forward_reads))
reverse_reads.cut <- file.path(path.cut, basename(reverse_reads))
R1.flags <- paste("-g", fwd_primer, "-a", rev_rc)
R2.flags <- paste("-G", rev_primer, "-A", fwd_rc)
# Run Cutadapt
for(i in seq_along(forward_reads)) {
  system2(cutadapt,args = c(R1.flags, R2.flags, "-n", 2, "-m", 1,
  # -n 2 required to remove FWD and REV from reads
  # -m 1 required to prevent error in PlotQualityProfile post cutadapt
  "-o", forward_reads.cut[i], "-p", reverse_reads.cut[i], # output files
  forward_reads.filtN[i], reverse_reads.filtN[i], "--discard-untrimmed",
  "--report=minimal"))
}

```

#### Search for primers & their reverse complements post-cutadapt
```{r post-cutadapt check}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = forward_reads.cut[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = reverse_reads.cut[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = forward_reads.cut[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = reverse_reads.cut[[1]]))

```

```{r Assess Raw Read Quality Post-cutadapt, eval=FALSE, include=FALSE}
#Display quality plot for forward read samples
forwardQual10_plot.cut <- plotQualityProfile(forward_reads.cut[1:10])
forwardQual20_plot.cut <- plotQualityProfile(forward_reads.cut[11:20])
forwardQual30_plot.cut <- plotQualityProfile(forward_reads.cut[21:30])
forwardQual40_plot.cut <- plotQualityProfile(forward_reads.cut[31:40])
forwardQual50_plot.cut <- plotQualityProfile(forward_reads.cut[41:50])
forwardQual60_plot.cut <- plotQualityProfile(forward_reads.cut[51:60])
forwardQual70_plot.cut <- plotQualityProfile(forward_reads.cut[61:70])
forwardQual80_plot.cut <- plotQualityProfile(forward_reads.cut[71:80])
forwardQual90_plot.cut <- plotQualityProfile(forward_reads.cut[81:90])
forwardQual100_plot.cut <- plotQualityProfile(forward_reads.cut[91:100])
forwardQual110_plot.cut <- plotQualityProfile(forward_reads.cut[101:110])
forwardQual117_plot.cut <- plotQualityProfile(forward_reads.cut[111:117])

#Display quality plot for  reverse read samples
reverseQual10_plot.cut <- plotQualityProfile(reverse_reads.cut[1:10])
reverseQual20_plot.cut <- plotQualityProfile(reverse_reads.cut[11:20])
reverseQual30_plot.cut <- plotQualityProfile(reverse_reads.cut[21:30])
reverseQual40_plot.cut <- plotQualityProfile(reverse_reads.cut[31:40])
reverseQual50_plot.cut <- plotQualityProfile(reverse_reads.cut[41:50])
reverseQual60_plot.cut <- plotQualityProfile(reverse_reads.cut[51:60])
reverseQual70_plot.cut <- plotQualityProfile(reverse_reads.cut[61:70])
reverseQual80_plot.cut <- plotQualityProfile(reverse_reads.cut[71:80])
reverseQual90_plot.cut <- plotQualityProfile(reverse_reads.cut[81:90])
reverseQual100_plot.cut <- plotQualityProfile(reverse_reads.cut[91:100])
reverseQual110_plot.cut <- plotQualityProfile(reverse_reads.cut[101:110])
reverseQual117_plot.cut <- plotQualityProfile(reverse_reads.cut[111:117])
```
The primer-free sequence files are now ready to be analyzed through the DADA2 pipeline!

### Filter amd Trim Reads
```{r Filter and Trim}
filtered_out <- filterAndTrim(forward_reads.cut, filtered_forward_reads, reverse_reads.cut, filtered_reverse_reads, maxN = 0, maxEE = c(2,2), truncQ = 2, rm.phix = TRUE, compress = TRUE, multithread = TRUE)

head(filtered_out)

```

#### Quality plot of trimmed & filtered reads
```{r Assess Raw Read Quality Post-cutadapt, eval=FALSE, include=FALSE}
#Display quality plot for forward read samples
forwardQual10_plot.filt <- plotQualityProfile(filtered_forward_reads[1:10])
forwardQual20_plot.filt <- plotQualityProfile(filtered_forward_reads[11:20])
forwardQual30_plot.filt <- plotQualityProfile(filtered_forward_reads[21:30])
forwardQual40_plot.filt <- plotQualityProfile(filtered_forward_reads[31:40])
forwardQual50_plot.filt <- plotQualityProfile(filtered_forward_reads[41:50])
forwardQual60_plot.filt <- plotQualityProfile(filtered_forward_reads[51:60])
forwardQual70_plot.filt <- plotQualityProfile(filtered_forward_reads[61:70])
forwardQual80_plot.filt <- plotQualityProfile(filtered_forward_reads[71:80])
forwardQual90_plot.filt <- plotQualityProfile(filtered_forward_reads[81:90])
forwardQual100_plot.filt <- plotQualityProfile(filtered_forward_reads[91:100])
forwardQual110_plot.filt <- plotQualityProfile(filtered_forward_reads[101:110])
forwardQual117_plot.filt <- plotQualityProfile(filtered_forward_reads[111:117])

#Display quality plot for  reverse read samples
reverseQual10_plot.filt <- plotQualityProfile(filtered_reverse_reads[1:10])
reverseQual20_plot.filt <- plotQualityProfile(filtered_reverse_reads[11:20])
reverseQual30_plot.filt <- plotQualityProfile(filtered_reverse_reads[21:30])
reverseQual40_plot.filt <- plotQualityProfile(filtered_reverse_reads[31:40])
reverseQual50_plot.filt <- plotQualityProfile(filtered_reverse_reads[41:50])
reverseQual60_plot.filt <- plotQualityProfile(filtered_reverse_reads[51:60])
reverseQual70_plot.filt <- plotQualityProfile(filtered_reverse_reads[61:70])
reverseQual80_plot.filt <- plotQualityProfile(filtered_reverse_reads[71:80])
reverseQual90_plot.filt <- plotQualityProfile(filtered_reverse_reads[81:90])
reverseQual100_plot.filt <- plotQualityProfile(filtered_reverse_reads[91:100])
reverseQual110_plot.filt <- plotQualityProfile(filtered_reverse_reads[101:110])
reverseQual117_plot.filt <- plotQualityProfile(filtered_reverse_reads[111:117])
```

### Generate an Error Model
```{r learn errors}
err_forward_reads <- learnErrors(filtered_forward_reads, multithread = TRUE)
err_reverse_reads <- learnErrors(filtered_reverse_reads, multithread = TRUE)
error_model_forward <- plotErrors(err_forward_reads, nominalQ = TRUE)
error_model_forward
error_model_reverse <- plotErrors(err_reverse_reads, nominalQ = TRUE)
error_model_reverse
```

### Infer Amplicon Sequence Variants (ASVs) from the forward and reverse reads
```{r DADA2}
# Run DADA2 on forward reads
dada_forward <- dada(filtered_forward_reads, err = err_forward_reads, multithread = TRUE)
head(dada_forward)

# Run DADA2 on reverse reads
dada_reverse <- dada(filtered_reverse_reads, err = err_reverse_reads, multithread = TRUE)
head(dada_reverse)
```
### Merge the forward and reverse ASVs
```{r Merge}
# Merge forward ASVs and reverse ASVs
merged_amplicons <- mergePairs(dadaF = dada_forward, 
                               derepF = filtered_forward_reads, 
                               dadaR = dada_reverse, 
                               derepR = filtered_reverse_reads,
                               verbose = TRUE)
#Inspect the output
head(merged_amplicons)

```

### Generate an ASV count table
```{r seqtab}
seqtab <- makeSequenceTable(samples = merged_amplicons)

# size of count table
dim(seqtab)

# distribution of sequence lengths of all ASVs
hist(nchar(getSequences(seqtab)), main = "Length distribution of ASVs", 
     xlab = "ASV length (bp)")

```

### Search and Remove Chimeras
```{r remove chimeras}
seqtab_nochim <- removeBimeraDenovo(seqtab, verbose = TRUE, multithread = TRUE)

# percentage of chimeras removed
chim_removed <- (1-sum(seqtab_nochim)/sum(seqtab))*100

```

Chimeras represented `r chim_removed`% of merged reads.
### Tracking changes in read counts throughout the workflow
```{r track changes}
# A function to count number of reads
getN <- function(x) sum(getUniques(x))

# A table to track read counts
track <- cbind(filtered_out, sapply(dada_forward, getN),
               sapply(dada_reverse, getN),
               sapply(merged_amplicons, getN),
               rowSums(seqtab_nochim))
# Rename columns and rows
colnames(track) <- c("input","filtered","denoisedF","denoisedR","merged","chimera_free")
rownames(track) <- samples

head(track)
# Alternatively, we can look at the percentage of reads lost during each step
track_pct <- track %>% as.data.frame() %>% rownames_to_column(var = "sample") %>%
  rowwise() %>% mutate(filtered = filtered/input, denoisedF = denoisedF/input,
                       denoisedR = denoisedR/input, merged = merged/input, chimera_free = chimera_free/input)
head(track_pct)

# Display these changes as a line plot

track_plot <- track %>% as.data.frame() %>% rownames_to_column(var = "samples") %>% pivot_longer(input:chimera_free, names_to = "read_type", values_to = "num_reads") %>% mutate(read_type = fct_relevel(read_type, "input", "filtered", "denoisedF", "denoisedR","merged","chimera_free")) %>% ggplot(aes(x = read_type, y = num_reads, fill = read_type)) + geom_line(aes(group = samples), color = "grey") + geom_point(shape = 21, size = 3, alpha = 0.8) + scale_fill_brewer(palette = "RdYlGn") + theme_bw()+ labs(x = "Filtering Step", y = "Number of sequences") + theme(legend.position = "bottom", legend.title = element_blank(), axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))

track_plot
```
